### PR TITLE
shift left overflow check now done in goto_check

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6e68a12ac468d121c3f730494a49715c7e9fb271).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96221565
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: regression/cbmc/Overflow_Leftshift1/test.desc [hunk 1]
```diff
@@ -3,8 +3,11 @@ main.c
 --signed-overflow-check
 ^EXIT=10$
 ^SIGNAL=0$
-^\[.*\] .* arithmetic overflow on signed shl in .*: FAILURE$
-^\*\* 2 of 4 failed
+^\[.*\] line 6 arithmetic overflow on signed shl in .*: FAILURE$
+^\[.*\] line 9 arithmetic overflow on signed shl in .*: SUCCESS$
```
Comment (smowton) [inline regression/cbmc/Overflow_Leftshift1/test.desc:7]: Perhaps add a must-not-match for "line 12 arithmetic overflow" if that's intentionally not tested?

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -589,6 +589,74 @@ void goto_checkt::integer_overflow_check(
...
+      // yet this isn't an overflow
+      exprt neg_dist_shift;
+
+      if(distance_type.id() == ID_unsignedbv)
+        neg_dist_shift = false_exprt();
+      else
+        neg_dist_shift =
+          binary_relation_exprt(op, ID_lt, from_integer(0, distance_type));
+
+      // an undefined shift of a non-zero value always results in overflow
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: `an undefined shift of a non-zero value` --> `shifting a non-zero value by more than its width`?

File: src/analyses/goto_check.cpp [hunk 3]
```diff
@@ -589,6 +589,74 @@ void goto_checkt::integer_overflow_check(
...
+      const exprt undef = binary_predicate_exprt(
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: `undef` is defined but not used, so this probably doesn't achieve the "always overflow" behaviour the comment asks for

File: src/analyses/goto_check.cpp [hunk 4]
```diff
@@ -589,6 +589,74 @@ void goto_checkt::integer_overflow_check(
...
+
+      auto new_type = to_bitvector_type(op_type);
+      new_type.set_width(op_width * 2);
+
+      const exprt op_ext = typecast_exprt(op, new_type);
+
+      const exprt op_ext_shifted = shl_exprt(op_ext, distance);
+
+      // get top bits of the shifted operand
+      const exprt top_bits = extractbits_exprt(
```
Comment (smowton) [inline src/analyses/goto_check.cpp:638]: ? Why shift and then extract? All the bits we're interested in testing are present in the original value.

**Review by smowton — CHANGES_REQUESTED** [review]
At least part of the implementation looks to be mistaken

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: Fixed

File: src/analyses/goto_check.cpp (hunk 3 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: Indeed, fixed

File: src/analyses/goto_check.cpp (hunk 4 above)
Comment (kroening) [reply src/analyses/goto_check.cpp:638]: Our extactbits matches the one in the SMT-LIB, and thus, only does constants.

File: regression/cbmc/Overflow_Leftshift1/test.desc (hunk 1 above)
Comment (kroening) [reply regression/cbmc/Overflow_Leftshift1/test.desc:7]: ok

**Review by smowton — COMMENTED** [review]
Thanks for the fixes, looks good except needs a case in the test for shifting more than the type width?

**Comment by kroening** [comment]
ok, added

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 444666753ef8d0c3bf32ba7695764dca436c3f64).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96338889
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3be9b5efc5f1af90d2af0ed688b45b11c8848dde).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96339556

**Review by smowton — APPROVED** [review]

**Pull request merged**